### PR TITLE
[SPARK-25828][K8S] Bumping Kubernetes-Client version to 4.1.0

### PR DESCRIPTION
**Pull request by ifilonenko**
## What changes were proposed in this pull request?

Changed the `kubernetes-client` version and refactored code that broke as a result

## How was this patch tested?

Unit and Integration tests

### CODE REVIEW

File: resource-managers/kubernetes/core/pom.xml [hunk 1]
```diff
@@ -29,7 +29,7 @@
   <name>Spark Project Kubernetes</name>
   <properties>
     <sbt.project.name>kubernetes</sbt.project.name>
-    <kubernetes.client.version>3.0.0</kubernetes.client.version>
+    <kubernetes.client.version>4.0.0</kubernetes.client.version>
```
Comment (skonto) [inline resource-managers/kubernetes/core/pom.xml]: 4.1.0?

Comment (ifilonenko) [reply resource-managers/kubernetes/core/pom.xml]: hmm, agreed. Maybe 4.1.0 is a more stable version

Comment (skonto) [reply resource-managers/kubernetes/core/pom.xml]: Yeah just wondering I don't know. What matters is what each version supports. Still no version supports 1.11 https://github.com/fabric8io/kubernetes-client/issues/1235

**Review by felixcheung — COMMENTED** [review]
should there be some doc changes on supported client version

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala [hunk 2]
```diff
@@ -157,7 +157,10 @@ private[spark] object KubernetesUtils {
       }.getOrElse(Seq(("container state", "N/A")))
   }
 
-  def formatTime(time: Time): String = {
-    if (time != null) time.getTime else "N/A"
+  def formatTime(time: String): String = {
```
[the remark this answers is not in the document]
Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: The `time.getTime` is now a String. I wanted to keep the functionality, I would remove if you deem it to be unnecessary.

**Review by liyinan926 — COMMENTED** [review]
LGTM. But looks like the build failed on dependency tests. Please take a look.

**Comment by ifilonenko** [comment]
In reference to manifests that need to be updated in `dev/deps/spark-deps-hadoop-2.7` and `dev/deps/spark-deps-hadoop-3.1`. This most recent commit should fix

**Comment by ifilonenko** [comment]
retest this please

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala [hunk 3]
```diff
@@ -157,7 +157,9 @@ private[spark] object KubernetesUtils {
...
+    // TODO: Investigate form `time` comes in as there was a DataType
+    // change from Time to String
```
Comment (rvesse) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: The string returned is ISO8601 format so no need to do anything special here

**Review by rvesse — APPROVED** [review]
+1 Would like to get this in ASAP so I can rebase PR #22805 and avoid mixing unrelated changes into the integration testing enhancement PR per @liyinan926's request

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala (hunk 3 above)
Comment (rvesse) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: e.g. a sample from the logs from my PR where I had made a similar change:

```
creation time: 2018-10-24T17:17:03Z
```

**Comment by ifilonenko** [comment]
@felixcheung and @erikerlandson for merge

**Comment by erikerlandson** [comment]
merging

**Pull request closed without merging**